src/trie.js

```diff
@@ -0,0 +1,97 @@
+const _ = require('lodash');
+const {scopeCompare} = require('taskcluster-lib-scopes');
+
+/**
+ * Build a trie to expand patterns involving kleene stars.
+ *
+ * Given input of the form [{pattern, scope}], this builds a trie
+ * of states of the form:
+ * ```js
+ * {
+ *   'scopes': [..],      // scopes to add when entering this state
+ *   'a': { ... },        // next state if the current character is 'a'
+ *   ...
+ *   'end': { ... },      // state if this is the end of the input
+ * ```
+ *
+ * This takes into account both scope expansion (kleene star in the input
+ * to the trie) and role expansion (kleene star in the pattern).  The
+ * resulting trie does *not* take into account recursive application of
+ * patterns.
+ *
+ * Note that, as applied to roles, the PATTERN for every rule begins with
+ * `assume:`, but nothing in this file is specific to that prefix.
+ */
+exports.generateTrie = (rules) => {
+  // generate the state for rules i .. n-1 starting at character position k.
+  // This assumes that the rules have the same value at positions < k, as
+  // assured by the sort order.  Thus we have three segments to work with,
+  // in order:
+  //         k
+  // i -> ...*   -- (A) terminal * in pattern
+  //      ...    -- (B) end of pattern (where pattern[k-1] != '*')
+  //      ...*a  \
+  //      ...a   -- (C) nonterminal * or other characters in pattern
+  //      ...ab  /
+  // n ->
+  let gen = (i, n, k) => {
+    let state = {};
+
+    let j = i;
+    while (j < n) {
+      // find the current segment, at indices [seg .. j-1]
+      let seg = j;
+      let current = rules[j].pattern[k]; // current character
+      if (current === '*' && rules[j].pattern.length === k + 1) {
+        // segment type A, terminal * in pattern
+        do {
+          j++;
+        } while (j < n && rules[j].pattern[k] == '*' && rules[j].pattern.length === k + 1);
```

Comment:
This part is necessary if you have multiple rules on the form `{pattern, scope}` where `pattern` is the same string.
You cannot possibly have that.

1. it's not possible with how we define roles
2. if we did have this we should just merge them into a single `{pattern, scopes}` rule, by merging `scopes`
3. this is never generated form `buildResolver`

I assume this is a left over from when you had `{pattern, scope}` instead of `{pattern, scopes}`. We can get some simplicity of case (A) and (B), but not doing this... simply requiring pattern uniqueness for input.